ucsb-cs48-s20/project-s1-t4-missile
# Chat

Closes #4 

In this PR:

- Added a chat system so that players can chat with each other
- Players can type messages into the input box and click send to send to all other players in the game
- Players are notified through the chat when other players join and leave the game
- The chat area can scroll to display more messages when messages fill past the bounds of the chat
- When a player has scrolled up in the chat, they can click a button to automatically scroll to the bottom of the chat
- Players can send emojis

To test:

- When you initially join the game as player 1, you will receive a welcome message in the chat
- When you join the game in another tab as player 2, player 1 will be notified through the chat that player 2 has joined
- When you type something in player 1's chat input box and click send:
    - Player 1 will see the message appear on the right side of the chat area in a blue container
    - Player 1's chat input box will be cleared
    - Player 2 will see the message appear on the left side of the chat area in a light gray container
- When you close player 2's tab, player 1 will be notified through the chat that player 2 has left
- When you send enough messages to fill the chat area, a scrolling bar will appear that allows you to scroll up and down the chat
- When you scroll up the chat, a gray button will appear on the bottom right that lets you instantly scroll back to the bottom of the chat
- If you are at the bottom of the chat when a new message appears in the chat area, the chat area will autoscroll to show the new message
- If you are _not_ at the bottom of the chat when a new message appears, the chat will _not_ autoscroll
- When you type an emoji such as `:)` into the chat, it will be displayed as the actual emoji
